craft_parts/sources/snap_source.py

```diff
@@ -0,0 +1,131 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The snap source handler."""
+
+import os
+import shutil
+import tempfile
+from typing import Optional
+
+import yaml
+
+from craft_parts.dirs import ProjectDirs
+from craft_parts.utils import file_utils
+
+from . import errors
+from .base import FileSourceHandler
+
+
+class SnapSource(FileSourceHandler):
+    """Handles downloading and extractions for a snap source.
+
+    On provision, the meta directory is renamed to meta.<snap-name> and, if present,
+    the same applies for the snap directory which shall be renamed to
+    snap.<snap-name>.
+    """
+
+    def __init__(
+        self,
+        source,
+        source_dir,
+        *,
+        application_name: str = None,
+        source_tag: str = None,
+        source_commit: str = None,
+        source_branch: str = None,
+        source_depth: Optional[int] = None,
+        source_checksum: str = None,
+        project_dirs: ProjectDirs = None,
+    ) -> None:
+        super().__init__(
+            source,
+            source_dir,
+            application_name=application_name,
+            source_tag=source_tag,
+            source_commit=source_commit,
+            source_branch=source_branch,
+            source_depth=source_depth,
+            source_checksum=source_checksum,
+            project_dirs=project_dirs,
+            command="unsquashfs",
+        )
+
+        if source_tag:
+            raise errors.InvalidSourceOption(source_type="snap", option="source-tag")
+
+        if source_commit:
+            raise errors.InvalidSourceOption(source_type="snap", option="source-commit")
+
+        if source_branch:
+            raise errors.InvalidSourceOption(source_type="snap", option="source-branch")
```

Comment:
IMO this complexity is not worth it. Why not just leave these parameters out of `__init__` signature?

If the developer is instantiating this class with the wrong parameters, it will fail anyway (maybe not with a super polished message, but that's for the developer, not the end user).

If this super important, though, shouldn't be tested?

Reply:
And why only `source-tag`, `source-commit` and `source-branch` are tested, but not `source-depth`? Will add a note to the source code to refactor and/or add tests in a follow-up PR.

Reply:
Thanks!